docs/LOGGING.md

```diff
@@ -30,28 +30,49 @@ Logs coming from xen-tools container are of three kinds:
+"loguploader" is a pillar service which is responsible for upload the gzip log files to the cloud server. The binary data of a gzip file is the payload portion of the authentication protobuf envolope structure. This is similar to all the other EVE POST messages, except that in those messsages the payload usually is data of another protobuf structure.
+
+The upload is one gzip file at a time. The "loguploader" finds the earliest timestamp from the gzip file's filename and sends the data to the cloud server. If the upload is sucessful, then the uploaded gzip file is removed from the directory. If the upload encounters an error, it will come back to retry again. There can be several different failure cases:
+
+1) the upload has no reply from the server and is TCP timed out
+2) the upload gets http status code of 503
+3) the upload gets http status code of 4xx
+4) the upload gets other types of error
+
+For the case 1 and 2, "loguploader" will just come back to retry the same gzip file again (the earliest in timestamp). After it repeats this 3 times, the "uploader" will mark the cloud server as "unreachable". This "unreachable" status affects the disk space management by "newlogd" as mentioned above. For case 3, if it repeats continously for 10 times for the same gzip file, this gzip file will be moved into "/persist/newlog/failedUpload" directory, this is to prevent one bad file stops the other log files to be uploaded forever. There is no special action for case 4, the "loguploader" will come back to retry again later.
+
+The uploading is controlled on a scheduled timer. When the timer fires, the "loguploader" checks both "devUpload" and "appUpload" directories, and picks file the earliest in timestamp of the gzip filename in the directory for uploading. The duration of the timer delay stays the same for 30 minutes, then it is recalculated. The current EVE implementaion calculates the timer delay based on those conditions:
+
+* if the cloud server is in "unreachable" status, then the delay is set to a random in the range of 3 minutes to 15 minutes. This is mainly due to if the cloud server is out of reach for a while, many devices has many accumulated logs to be sent to the cloud, to space out the load in a longer duration will help the server side dealing with huge load in the initial startup stage.
+* the device boots up, the timer value is set to 90 seconds (the first 30 minutes)
+* the total number of gzip log files currently remains in the directories (the below timer value with 15% randomness):
+  * 90 seconds if total file number is less than 5
+  * 45 seconds if total file number is less than 25
+  * 30 seconds if total file number is less than 50
+  * 15 seconds if total file number is less than 200
+  * 8 seconds if total file number is more than 200
```

Comment:
Does this mean that even if the controller and network are super fast you will be sending at most one file every 8 seconds? Or is 8 seconds the spacing if there are errors?
Does the 8 seconds count from the start or finish of the previous file?

Reply:
currently I don't take cpu/bandwidth/error into consideration. it's only based on file numbers left in the directory. we can change anytime later. the timer is started after finish the last one, and those timers are randomized for 15%

Reply:
unless if we fail to upload, then after 3 continuous failure, that will be a slow timer, but still follows previous half an hour calculation

Reply:
But there is always at least an 8 second gap from sending one file until you start sending the next file?

Reply:
correct. my testings with full debug, this 8 seconds seems to be fine, it does not lag behind of log files.

Reply:
My only question is if running with debug and there is a network outage and there is plenty of free bandwith whether there is a case when one could want less than 8 and go as low as zero.

But we can keep it this way; I suspect we might later need to add some average log rate limit (measured in bytes per month) and when we do we can have these timers take that into account.

Reply:
ok, for now. I'm sure we'll fine tune this with more experience.
What I have noticed with the above schedule is that, when there is lots of files, the 8 seconds kicks in, and soon the file number reduces. For every half an hour re-evaluation, we noticed that the numbers is below say 50, then the next half an hour will be using the 30 seconds timer. the end result is that, it seems the file numbers remain in the directory is hoovering around 50 files number.
So I'll do minor change to the above the, when re-evaluation every 30 min, the timer can only adjust downwards unless the file number is below 5.